internal/server/providers/azure_test.go

```diff
@@ -0,0 +1,328 @@
+			},
+		},
+		{
+			name: "failure to sync groups does fail sync",
+			setupFunc: func(t *testing.T) *models.Identity {
+				graphGroupMemberEndpoint = "https://" + serverURL + "/v1.0/me/memberOf/fail"
+
+				user := &models.Identity{
+					Name: "nwheeler@example.com",
+				}
+
+				err = data.CreateIdentity(db, user)
+				assert.NilError(t, err)
+
+				pu := &models.ProviderUser{
+					ProviderID: provider.ID,
+					IdentityID: user.ID,
+
+					Email:        user.Name,
+					RedirectURL:  "http://example.com",
+					AccessToken:  models.EncryptedAtRest("aaa"), // this is used to fail the groups call, in reality this token should be valid
+					RefreshToken: models.EncryptedAtRest("bbb"),
+					ExpiresAt:    time.Now().Add(time.Minute * -5),
+					LastUpdate:   time.Now(),
+				}
+
+				err = data.UpdateProviderUser(db, pu)
+				assert.NilError(t, err)
+
+				return user
+			},
+			infoResponse: `{
+				"sub": "o_aaabbbccc",
+				"sub": "o_aaabbbccc",
+				"name": "Nancy Wheeler",
+				"family_name": "Wheeler",
+				"given_name": "Nancy",
+				"picture": "https://graph.microsoft.com/v1.0/me/photo/$value"
+			}`,
+			verifyFunc: func(t *testing.T, err error, user *models.Identity) {
+				assert.NilError(t, err)
+			},
+		},
+		{
+			name: "groups are set from graph response",
+			setupFunc: func(t *testing.T) *models.Identity {
+				graphGroupMemberEndpoint = "https://" + serverURL + "/v1.0/me/memberOf"
+
+				user := &models.Identity{
+					Name: "jhopper@example.com",
+				}
+
+				err = data.CreateIdentity(db, user)
+				assert.NilError(t, err)
+
+				pu := &models.ProviderUser{
+					ProviderID: provider.ID,
+					IdentityID: user.ID,
+
+					Email:        user.Name,
+					RedirectURL:  "http://example.com",
+					AccessToken:  models.EncryptedAtRest("aaa"), // this is used to fail the groups call, in reality this token should be valid
+					RefreshToken: models.EncryptedAtRest("bbb"),
+					ExpiresAt:    time.Now().Add(time.Minute * -5),
+					LastUpdate:   time.Now(),
+				}
+
+				err = data.UpdateProviderUser(db, pu)
+				assert.NilError(t, err)
+
+				return user
+			},
+			infoResponse: `{
+				"sub": "o_aaabbbccc",
+				"sub": "o_aaabbbccc",
+				"name": "Jim Hopper",
+				"family_name": "Hopper",
+				"given_name": "Jim",
+				"picture": "https://graph.microsoft.com/v1.0/me/photo/$value"
+			}`,
+			verifyFunc: func(t *testing.T, err error, user *models.Identity) {
+				assert.NilError(t, err)
+
+				pu, err := data.GetProviderUser(db, provider.ID, user.ID)
+				assert.Assert(t, err)
+
+				assert.Assert(t, len(pu.Groups) == 2)
+
+				groups := make(map[string]bool)
+				for _, g := range pu.Groups {
+					groups[g] = true
+				}
+
+				assert.Assert(t, groups["Everyone"])
+				assert.Assert(t, groups["Developers"])
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			user := test.setupFunc(t)
+			server.userInfoResponse = test.infoResponse
```

Comment:
(Not a blocker) Since this modifies the test fixture used by all test cases it might make sense to move `setupOIDCTest` and anything that depends on the server into this run function. It seems like all of that runs pretty fast, so we don't pay much of a penalty for repeating it, and that might prevent surprising cases where the server state is saved between tests.

That would make the `assert.Check` suggestion for errors better, because then the `testing.T` passed to the `server.run` would be for only the test case, not the full test function.